Build Foreman with ansible-container and deploy on Openshift V3

ehelms opened the pull request:
This work is a work in progress to explore the ability to deploy Foreman and Katello on to an OpenShift V3 (container and Kubernetes based environment) environment to explore harnessing the power of the OpenShift platform to manage the application and scale out. The `ansible-container` project is used to build the containers instead of traditional Dockerfiiles. This allows building containers using Ansible and to run them both locally as well as deploying them to Openshift.

For more detailed information on this deployment, how to use, test and contribute see the README (https://github.com/ehelms/forklift/blob/openshift-2.0/openshift/README.md).

This work combines work built on top of the foreman-docker-compose repository (https://github.com/shlomizadok/foreman-docker-compose) along with previous work by the Pulp and Candlepin teams to build out images for use on OpenShift. This also pulls in a starter repo (https://github.com/ehelms/foreman-openshift) that was used to hold initial work on the OpenShift piece.

I have initially pulled this work in Forklift for the following:

 * Forklift's goal is around putting Foreman into user and developer environments
 * Forklift has a broader audience than a new set of repositories and branches would have
 * To get more eyes and interest on this idea for deploying Foreman
 * If others feel this would be better maintained in a separate repository, I'm open to that discussion with the hope that the previous reasons will be thoughtfully considered with respect on how to position this.

ehelms commented:
This may or may not depend on the outcome of https://github.com/theforeman/forklift/pull/379 to go in. This is already isolated though does have some redundancy due to the path lookups by `ansible-container`. 

I would also appreciate some input of how to pull this work in. Currently, the README has a list of TODOs that I would prefer to move to proper issues marked against either a project or milestone in Github for visibility.

ehelms commented:
Also, there is a general architecture and breakdown of services present at:

https://github.com/ehelms/forklift/blob/5f39dc1a838654faac5f7b4877302d8b5724c2de/openshift/docs/architecture.png

ekohl commented:
That's a very useful architecture image, also just for the installer + puppet modules.

ehelms commented:
@ekohl I once set about to make a diagram of current architecture, ports and specifically SSL expectations. This is probably slightly out of date but it could be useful to you -- https://github.com/ehelms/connection_diagram/blob/master/katello.png

ehelms commented:
Recent Updates:
  * Moved Foreman tasks and dynflow to their own service
  * Updated architecture diagram to reflect foreman tasks on its own
  * Added an architecture documentation page to outline each of the services in some detail

ehelms commented:
Latest updates, I am working on updating the build strategy to use the latest ansible-container (0.9) which requires a bit of a migration.

ehelms commented:
Pushed an update that moves to using `ansible-container` 0.9+. There is a current issue affecting the ability to then push to Openshift:

https://github.com/ansible/ansible-container/issues/542

ekohl commented on openshift/cluster-up.yml:
```diff
@@ -0,0 +1,52 @@
+  vars:
+    openshift_data_dir: /home/origin
+  tasks:
+    - name: 'Check  cluster status'
```
Nit: double space

ekohl commented on openshift/install-ansible-container.yml:
```diff
@@ -0,0 +1,18 @@
+
+    - name: 'Install ansible-container'
+      become: true
+      command: 'pip install -e .[docker,openshift]'
```
Would https://docs.ansible.com/ansible/pip_module.html work here?

ekohl commented on openshift/roles/candlepin/files/gen-certs:
```diff
@@ -0,0 +1,50 @@
+
+shift $(($OPTIND - 1))
+
+rpm -q openssl > /dev/null
```
```which openssl``` would be generic

ekohl commented on openshift/roles/pulp/tasks/main.yml:
```diff
@@ -0,0 +1,69 @@
+
+- name: 'Add logs.py patch'
+  get_url:
+    url: https://raw.githubusercontent.com/ehelms/pulp/log-to-file/server/pulp/server/logs.py
```
Any plan to fix this properly?

ehelms replied:
A community user actually added this feature in Pulp as of 2.13 and since this pulls from nightly I could not work on updating this.

ekohl commented on openshift/test/ansible/roles/foreman_repositories/tasks/release_repos.yml:
```diff
@@ -0,0 +1,12 @@
+  yum:
+    name: http://fedorapeople.org/groups/katello/releases/yum/foreman/1.13/el{{ ansible_distribution_major_version }}/x86_64/foreman-release.rpm
+    state: present
+  when: foreman_repositories_version == 1.13 and ansible_distribution_major_version == "6"
```
Do you really want to support 1.13 on EL6 on openshift?

ehelms replied:
Haha no -- this was copied from the foreman_repositories role that I think still exists with this in it today.

ekohl commented on openshift/test/ansible/roles/katello_repositories/tasks/main.yml:
```diff
@@ -0,0 +1,9 @@
+  when: katello_repositories_use_koji
+
+- include: qpid.yml
+  when: ansible_distribution_major_version == "6" and katello_repositories_version == 3.2
```
Do you really want to support this?

ehelms replied:
See above comment :)

ehelms commented:
Latest round of updates:

  * Working on adding secrets support to ansible-container (https://github.com/ansible/ansible-container/pull/612) to allow distribution of certificates and passwords
  * Added foreman-memcache plugin and memcache container, setting replicas to 2 by default for Foreman for HA testing
  * Add certificate generation role, certificate generation playbook to generate an Ansible variable file of certificates that could be used directly or encrypter into vault

Using the current state of the PR, including my PR to ansible-container, I've been able to:

 * Create Orgs and Locs
 * Create products and repositories
 * Sync a repository
 * View the contents of the repository on the content server service
 * Create and publish content views

ares commented:
I looked at the [diagram](https://raw.githubusercontent.com/ehelms/forklift/5f39dc1a838654faac5f7b4877302d8b5724c2de/openshift/docs/architecture.png), just note that also foreman-proxy might need a storage. It can run dynflow, and at least openscap plugin stores arf_reports there. Otherwise looks good.

ehelms commented:
@ares When it's running dynflow, what backing storage is it using? Sqlite database? Openscap would just need a mounted storage for saving files, not a database service running?

ares commented:
correct, it's using sqlite DB which is stored at /var/lib/foreman-proxy/dynflow/dynflow.sqlite (by default). openscap plugin stores only plain files in /var/lib/foreman-proxy/openscap/{content,reports,failed}

ohadlevy commented on containers/docs/setup-and-install.md:
```diff
@@ -0,0 +1,130 @@
+
+This project is currently using the bleeding edge version of `ansible-container` and will be installed from source. To simplify install, an Ansible playbook has been provided to do the installation:
+
+    ansible-playbook install-ansible-container.yml
```
should mention to change to containers subfolder

ohadlevy replied:
when running i'm getting forklift/containers/.vault_pass was not found

ohadlevy replied:
i also needed to install python3-devel to get it running

ehelms replied:
I added a message above this now about this requirement.

ohadlevy commented on containers/README.md:
not relevant bz?

ekohl commented on containers/ansible.cfg:
```diff
@@ -0,0 +1,4 @@
+[defaults]
+vault_password_file = .vault_pass
+retry_files_enabled = False
+roles_path=roles:$PWD/../playbooks/roles:ansible-deployment/roles
```
If this is supposed to refer to the main directory then it needs to drop the playbooks.

ekohl commented on containers/install-ansible-container.yml:
```diff
@@ -0,0 +1,20 @@
+        update: yes
+
+    - name: 'Install ansible-container'
+      become: true
```
You could use ```pip3 install --user --editable .[docker,openshift]``` and drop the become: true. You do need to be aware that it places the binary in ```~/.local/bin``` which may not be in your ```$PATH```. Maybe you could also use ```.tmp/ansible-container``` instead of ```.``` and drop the ```chdir```.

ehelms commented on containers/README.md:
Oops, copy/psate fail.

ehelms commented:
@ekohl I'd like to work towards merging this so that others can start to contribute and tackle different parts. What would you like to see done before this is merged?

ekohl commented on containers/docs/setup-and-install.md:
```diff
@@ -0,0 +1,132 @@
+
+Now build the containers (and grab a coffee):
+
+    ansible-container build
```
You should now refer to the ```build.sh``` to load the correct roles.

ekohl commented on containers/roles/candlepin/files/entrypoint.sh:
```diff
@@ -0,0 +1,25 @@
+
+/usr/bin/gen-certs
+
+liquibase \
```
In puppet we changed it to use cpdb.

ehelms replied:
Filed an issue for this

ekohl commented on containers/roles/foreman/files/entrypoint.sh:
```diff
@@ -0,0 +1,14 @@
+foreman-rake db:migrate
+echo $SEED_ADMIN_PASSWORD
+SEED_ADMIN_PASSWORD=$SEED_ADMIN_PASSWORD foreman-rake db:seed
+foreman-rake apipie:cache:index
```
I recall this step being very slow. Future work should have a look at improving this.

ehelms replied:
It is, there is a strategy for doing this in the build phase via a sqlite database. Would be awesome if we could eliminate the need for a DB.

ehelms replied:
Filed an issue for this

ekohl commented on containers/roles/foreman/files/wait_on_postgres.py:
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env python
```
A nice future improvement would be to have this in a common role so this code isn't duplicated.

ehelms replied:
Filed an issue for this

ekohl commented on containers/roles/foreman/tasks/install-puma.yml:
```diff
@@ -0,0 +1,22 @@
+    - libxslt-devel
+
+- name: 'Install puma-rails'
+  shell: scl enable tfm 'gem install puma-rails'
```
You have a TODO somewhere to package this? It'd be nice if in the future foreman containers don't need a development stack installed to keep the size down.

ehelms replied:
Filed an issue for this

ekohl commented on containers/test/ansible/container.yml:
```diff
@@ -0,0 +1,20 @@
+  bats:
+    from: centos:7
+    environment:
+      - FOREMAN_HOSTNAME=foreman-8080-foreman.192.168.1.33.xip.io
```
This feels fragile to me since this IP is defined nowhere else AFAIK.

ehelms replied:
Filed an issue for this

ekohl commented on containers/test/ansible/roles/epel_repositories/tasks/main.yml:
```diff
@@ -0,0 +1,3 @@
+---
```
Is it on purpose that these roles are duplicated?

ekohl commented:
It looks like I can build after a ```docker pull centos/mongodb-26-centos7:latest```

ekohl commented:
The next error is:

fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Configuration error - The Compose file '/tmp/ansiblemDUzBw/docker-compose.yml' is invalid because:\nUnsupported config option for services.candlepin: 'secrets'\nUnsupported config option for services.content-server: 'secrets'\nUnsupported config option for services.foreman: 'secrets'\nUnsupported config option for services.foreman-tasks: 'secrets'\nUnsupported config option for services.pulp: 'secrets'"}

ekohl commented on containers/docs/setup-and-install.md:
```diff
@@ -0,0 +1,132 @@
+
+Once the containers have been built locally into Docker, `ansible-container` can be used to run the entire stack locally:
+
+    ansible-container run
```
I now need to specify the ```-roles-path ../roles```

ehelms replied:
I should just rip this out for now. The `run` won't work currently due to port overlap and secrets support missing.

ehelms commented:
Strange that it did not pull that for you during build, I have not seen that issue. As for the futures and TODOs, we talked briefly I think in the past but I'll re-iterate my ideas for managing that to see your thoughts:

 1) Keep list of TODOs in the README (not my favorite)
 2) Create Github issues with a Github project dedicated to container TODOs
 3) Create Github issues with a Github milestone dedicated to container TODOs
 4) Manage through Redmine (not my favorite either since it separates the issues)

ehelms commented:
@ekohl Are you using the local ansible-container that is installed from my branch?

ekohl commented:
> Create Github issues with a Github project dedicated to container TODOs

This would be where I'd start as a general tracker. Maybe create a milestone as well so we can link PRs to it but that might just be overkill.

> @ekohl Are you using the local ansible-container that is installed from my branch?

I am

ehelms commented:
So I opened a project and a series of issues labeled for containerization. For some reason I can't drag'n'drop issues onto the Backlog column on the project though? Could you test @ekohl maybe its my browser?

What else would you like to see @ekohl ?

https://github.com/theforeman/forklift/projects/1
https://github.com/theforeman/forklift/issues?q=is%3Aopen+is%3Aissue+label%3AContainers

ekohl commented:
Other than making sure Rubocop ignores the .spec file I think this is good to go.

ehelms commented:
@ekohl rubocop updated and added a few extra travis lint checks

ehelms commented:
Ansible lint rules are now passing as well

ehelms commented:
@ekohl If you get a chance to re-look at this, and the issues made.
